Doc fuzzy gaussian (#5377)
qgis/QGIS-Documentation: docs/user_manual/processing_algs/qgis/rasteranalysis.rst

      :local:
      :depth: 1

.. _qgisfuzzifyrastergaussianmembership:

Fuzzify raster (gaussian membership)
------------------------------------

> Transforms an input raster to a fuzzified raster by assigning to each pixel

Reviewer: ```suggestion
Transforms an input raster to a fuzzified raster by assigning a membership
```

> membership values, following a gaussian membership function.

Reviewer: ```suggestion
value to each pixel, using a Gaussian membership function.
```

> Membership values ranges from 0 to 1. In the fuzzified raster, a value of 0 implies

Reviewer: ```suggestion
Membership values range from 0 to 1. In the fuzzified raster, a value of 0 implies
```

> no membership with the defined fuzzy set whereas a value of 1 means full membership.

Reviewer: ```suggestion
no membership of the defined fuzzy set, whereas a value of 1 means full membership.
```

> The gaussian membership function is defined as |gaussian_formula|, where f1 is the

Reviewer: ```suggestion
The gaussian membership function is defined as

|gaussian_formula|

where *f1* is the spread and *f2* the midpoint.
```

spread and f2 the midpoint.

.. figure:: img/gaussianimage.png
  :align: center
  
  Fuzzify raster example. Input raster source: Land Tirol - data.tirol.gv.at.

Parameters
..........

.. list-table::
   :header-rows: 1
   :widths: 20 20 20 40
   :stub-columns: 0

   * - Label
     - Name
     - Type
     - Description
   * - **Input Raster**
     - ``INPUT``
     - [raster]
     - Input raster layer
   * - **Band Number**
     - ``BAND`` 
     - [raster band]
       
       Default: The first band of the raster layer
     - If the raster is multiband, choose the band that
       you want to fuzzify.
   * - **Function midpoint**
     - ``FUZZYMIDPOINT``
     - [number]

       Default: 10
     - Midpoint of the gaussian function
   * - **Function spread**
     - ``FUZZYSPREAD``
     - [number]

       Default: 0.01
     - Value to use for nodata in the output layer
   * - **Fuzzified raster**
     - ``OUTPUT``
     - [same as input]

>      - Output raster layer containing the result

Reviewer: I forgot about this one.
```suggestion
     - Specification of the output raster. One of:

       * Save to a Temporary File
       * Save to File...

       The file encoding can also be changed here.
```

Outputs
.......

.. list-table::
   :header-rows: 1
   :widths: 20 20 20 40
   :stub-columns: 0

   * - Label
     - Name
     - Type
     - Description
   * - **Fuzzified raster**
     - ``OUTPUT``
     - [same as input]
     - Output raster layer containing the result

   * - **CRS authority identifier**
     - ``CRS_AUTHID``
     - [crs]
     - The coordinate reference system of the output raster layer
   * - **CRS authority identifier**
     - ``EXTENT``
     - [extent]
     - The spatial extent of the output raster layer
   * - **Width in pixels**
     - ``WIDTH_IN_PIXELS``
     - [integer]
     - The width in pixels of the output raster layer
   * - **Height in pixels**
     - ``HEIGHT_IN_PIXELS``
     - [integer]
     - The height in pixels of the output raster layer
   * - **Total pixel count**
     - ``TOTAL_PIXEL_COUNT``
     - [integer]
     - The count of pixels in the output raster layer

>    * - **Output**

Reviewer: `OUTPUT` has already been listed (as *Fuzzified raster*), so it should not be repeated here.
```suggestion
```

>      - ``OUTPUT``

Reviewer: ```suggestion
```

>      - [raster]

Reviewer: ```suggestion
```

>      - Output raster layer containing the result

Reviewer: ```suggestion
```

      :local:
      :depth: 1

.. _qgisfuzzifyrastergaussianmembership:

Fuzzify raster (gaussian membership)
------------------------------------

Transforms an input raster to a fuzzified raster by assigning a membership
value to each pixel, using a Gaussian membership function.
Membership values range from 0 to 1. In the fuzzified raster, a value of 0 implies
no membership of the defined fuzzy set, whereas a value of 1 means full membership.
The gaussian membership function is defined as

|gaussian_formula|

where *f1* is the spread and *f2* the midpoint.

> spread and f2 the midpoint.

Reviewer 2: should it be?

Author: Sorry, @DelazJ I just accepted all the suggesstions of @havatv without reading carefully. You are rigth. It is duplicated.

Reviewer: Sorry about that.

  * Variety
  * Variance
  * All

.. Substitutions definitions - AVOID EDITING PAST THIS LINE
   This will be automatically updated by the find_set_subst.py script.
   If you need to create a new substitution manually,
   please add it also to the substitutions.txt file in the
   source folder.

.. |gaussian_formula| image:: img/gaussianformula.png

>    :width: 8.3em

Reviewer: This one seems to become too big. And we have not used `:height:` in the QGIS documentation before. Is it necessary to use it here?

Author: Using only the width the formula become unnecessarily big.  I tried with the heigth and the proportion looks good. Also, because I wanted to have the formula inside the text. But, if is there another simple way, fine.

Reviewer 2: Sorry for being late but I very much prefer the formula inside the text: makes reading straightforward. and I did not check the rendering but yes, we do use :height: (https://github.com/qgis/QGIS-Documentation/blob/master/substitutions.txt) and in this case, i wonder if :width: is not unnecessary.

Reviewer 2: I guess there's no (easy?) way to write this formula in rst...

Author: > I guess there's no (easy?) way to write this formula in rst...

That would be nice. I tried a bit with :math but i got a warning
"WARNING: LaTeX command 'latex' cannot be run (needed for math display), check the imgmath_latex setting"
I do no have so much time to solve this. So, the img is fine for me.

Reviewer 2: Actually using the :math: role works, ie :math:`\mu(x)= e^{f_1*(x-f_2)^2}` (it complains about dvipng not installed on my system but after I installed it it works)
![image](https://user-images.githubusercontent.com/7983394/80796873-a22b5b00-8ba0-11ea-92f9-a5b226e39394.png)

Author: @DelazJ it looks nice. 
Perhaps @rduivenvoorde knows how to setup this in Windows?

Reviewer 2: > "WARNING: LaTeX command 'latex' cannot be run (needed for math display), check the imgmath_latex setting"

Nothing more in the message? Internet does not provide more details about this error, from a quick search. we already have the imgmath extension in our conf file.
Does this warning stop the build? It should normally keep going. And then how is the rendering in your local html? And you can push the suggested changes to the PR, so that we see what our CI infrastructure may say.

Author: ![image](https://user-images.githubusercontent.com/22812264/80806292-92b90b80-8bba-11ea-898d-74fe2af05ffb.png)
In this case, I just put next to the image (locally). 
The formula was added as suggested.

Reviewer 3: > Perhaps @rduivenvoorde knows how to setup this in Windows?
> update: The github editor does not show the formula properly either (I just tried),
> So I have no easy way to verify that I wrote the formula correctly. I will stick to the image, but feel free to edit the file and add the formula.

I'd stick with the image. Though maybe nicer it is probably a hassle to set latex up on Windows.
I'm not going to try...
And if this would mean that we can only build on Linux, we miss the boat again :-)
So please let's keep it easy for everybody to build html

Author: The image is back. Thanks @rduivenvoorde

Reviewer 2: OK. Let's stick to the image inline (size seems ok in the screenshot btw), it works on windows and github for everyone. Simpler and wiser.
Thanks

  * Variety
  * Variance
  * All

.. Substitutions definitions - AVOID EDITING PAST THIS LINE
   This will be automatically updated by the find_set_subst.py script.
   If you need to create a new substitution manually,
   please add it also to the substitutions.txt file in the
   source folder.

.. |gaussian_formula| image:: img/gaussianformula.png

>    :height: 1.4em

Reviewer: ```suggestion
   :height: 1em
```
If we are to keep it in-line, I still think it is too large.

Reviewer 2: @havatv Mind sharing the preview that bothers you, please? From what I see in rosa's screenshot, it does not look big to me. I would have suggested 1.5em and no white background but I was afraid I'd scare her so kept that for now.

Reviewer: ![image](https://user-images.githubusercontent.com/1598812/80978895-f182bd00-8e26-11ea-9937-c15c849fdad6.png)
Looks a bit excessive to me... Probably needs to be reduced more than to 1em.

Author: @havatv is this a preview in Github? 
See the image below with 0.5em,
![image](https://user-images.githubusercontent.com/22812264/80981375-2e03e800-8e2a-11ea-8283-f0945d2efe16.png)
it seems that changing the size here does not affect the rendering, I migth be wrong but I tried with different values.
Please check the preview I sent before (after compiling locally with 1.4em)

Reviewer 2: Isn't this a GitHub preview? Not a HTML rendering , which is our end product, and the one we should prioritarily care about.

Reviewer: Yes, this is a github preview.  I assumed that the HTML rendering would be the same. Apparently not. What does it look like with HTML?

Author: In this case, I just put the fromula next to the image (locally). Just look at the size of the image. Hth.
![image](https://user-images.githubusercontent.com/22812264/81049118-ddd66580-8ebd-11ea-9dd3-ce23a1643cc1.png)

Reviewer: Looks fine. Sorry about the noise!
